File: desktop/src/services/updater.ts
import { createSignal } from 'solid-js';

export interface UpdateInfo {
  version: string;
  notes: string | null;
  date: string | null;
}

interface TauriCore {
  invoke<T>(cmd: string, args?: Record<string, unknown>): Promise<T>;
}

interface TauriWindow extends Window {
  __TAURI__?: {
    core: TauriCore;
  };
}

const [updateAvailable, setUpdateAvailable] = createSignal<UpdateInfo | null>(null);
const [updateBannerVisible, setUpdateBannerVisible] = createSignal(false);
const [updateDownloading, setUpdateDownloading] = createSignal(false);
const [updateProgress, setUpdateProgress] = createSignal<string>('');

export { updateAvailable, updateBannerVisible, updateDownloading, updateProgress };

export async function checkForUpdates(): Promise<UpdateInfo | null> {
  if (typeof window === 'undefined') {
    return null;
  }
  const w = window as TauriWindow;
  if (!w.__TAURI__?.core) {
    return null;
  }
  try {
    const result = await w.__TAURI__.core.invoke<UpdateInfo | null>('check_for_updates');
    if (result) {
      setUpdateAvailable(result);
      setUpdateBannerVisible(true);
    }
    return result;
  } catch (err) {
    console.error('Failed to check for updates:', err);
    return null;
  }
}

export async function installUpdate(): Promise<boolean> {
  if (typeof window === 'undefined') {
    return false;
  }
  const w = window as TauriWindow;
  if (!w.__TAURI__?.core) {
    return false;
  }
  setUpdateDownloading(true);
  setUpdateProgress('Downloading update...');
  try {
    await w.__TAURI__.core.invoke('install_update');
    setUpdateDownloading(false);
    setUpdateBannerVisible(false);
    return true;
  } catch (err) {
    console.error('Failed to install update:', err);
    setUpdateDownloading(false);
    setUpdateProgress('');
    return false;
  }
}

export function dismissUpdateBanner(): void {
  setUpdateBannerVisible(false);
}

File: desktop/src-tauri/src/updater.rs
use serde::Serialize;

#[derive(Debug, Serialize)]
pub struct UpdateInfo {
    pub version: String,
    pub notes: Option<String>,
    pub date: Option<String>,
}

/// Check for available updates via the Tauri updater plugin.
/// Returns update info if an update is available, None otherwise.
#[tauri::command]
pub async fn check_for_updates(app: tauri::AppHandle) -> Result<Option<UpdateInfo>, String> {
    use tauri_plugin_updater::UpdaterExt;

    let updater = app
        .updater()
        .map_err(|e| format!("Updater not available: {}", e))?;

    match updater.check().await {
        Ok(Some(update)) => {
            let version = update.version.clone();
            let notes = update.body.clone();
            let date = update.date.map(|d| d.to_string());
            tracing::info!("Update available: {}", version);
            Ok(Some(UpdateInfo {
                version,
                notes,
                date,
            }))
        }
        Ok(None) => {
            tracing::info!("No updates available");
            Ok(None)
        }
        Err(e) => {
            tracing::error!("Failed to check for updates: {}", e);
            Err(format!("Failed to check for updates: {}", e))
        }
    }
}

/// Download and install the available update.
/// Returns an error string if no update is available or installation fails.
#[tauri::command]
pub async fn install_update(app: tauri::AppHandle) -> Result<(), String> {
    use tauri_plugin_updater::UpdaterExt;

    let updater = app
        .updater()
        .map_err(|e| format!("Updater not available: {}", e))?;

    let update = updater
        .check()
        .await
        .map_err(|e| format!("Failed to check for updates: {}", e))?
        .ok_or_else(|| "No update available to install".to_string())?;

    tracing::info!("Downloading and installing update: {}", update.version);

    update
        .download_and_install(
            |chunk_length, content_length| {
                tracing::debug!(
                    "Downloaded {} bytes (total: {:?})",
                    chunk_length,
                    content_length
                );
            },
            || {
                tracing::info!("Download finished, installing...");
            },
        )
        .await
        .map_err(|e| format!("Failed to install update: {}", e))?;

    Ok(())
}
